File: web/src/import/diagnostics.test.ts
import { describe, expect, it } from 'vitest'
import { diagnoseExport } from './diagnostics'
import { PARSER_VERSION } from './spotify-parser'
import { openZipArchive } from './zip-reader'
import { listFixtureCases, readExpected, readFixtureArchive, recordingArchive } from '../test/listening-export-fixtures'

describe('diagnoseExport', () => {
  it('lists sentinel files with byte sizes only and never reads them', async () => {
    const piiCase = listFixtureCases().find((fixture) => fixture.name === 'account-pii-present')!
    const archive = recordingArchive(await openZipArchive(readFixtureArchive(piiCase.name)))
    const diagnostics = await diagnoseExport(archive)
    expect(diagnostics.source).toBe('spotify_export')
    expect(diagnostics.parserVersion).toBe(PARSER_VERSION)
    const expected = readExpected(piiCase.name, 'default')
    for (const ignored of expected.inventory.ignored) {
      expect(diagnostics.files).toContainEqual({ path: ignored.path, bytes: ignored.bytes, rows: null, headers: null })
    }
    for (const sentinel of piiCase.sentinelPaths) expect(archive.readPaths).not.toContain(sentinel)
    expect(JSON.stringify(diagnostics)).not.toContain('DO-NOT-READ')
    expect(diagnostics.files.map((file) => file.path)).toEqual(
      [...expected.inventory.read.map((file) => file.path), ...expected.inventory.ignored.map((file) => file.path)].sort(),
    )
  })

  it('reports row counts and top-level key names for the files the package reads, never values', async () => {
    const archive = await openZipArchive(readFixtureArchive('account-pii-present'))
    const { files } = await diagnoseExport(archive)
    const expected = readExpected('account-pii-present', 'default')
    const library = files.find((file) => file.path.endsWith('/YourLibrary.json'))!
    const playlist = files.find((file) => file.path.endsWith('/Playlist1.json'))!
    expect(library.rows).toBe(expected.inventory.read.find((file) => file.path === library.path)?.rows)
    expect(library.headers).toEqual(expect.arrayContaining(['tracks', 'artists']))
    expect(playlist.rows).toBe(1)
    expect(playlist.headers).toEqual(['playlists'])
    const snapshot = expected.snapshot as { tracks: { title: string; artist: string }[]; playlists: { name: string }[] }
    const text = JSON.stringify(files)
    for (const track of snapshot.tracks) {
      expect(text).not.toContain(track.title)
      expect(text).not.toContain(track.artist)
    }
    for (const playlistRow of snapshot.playlists) expect(text).not.toContain(playlistRow.name)
  })

  it('marks a broken file with null rows and headers and keeps the valid one', async () => {
    const { files } = await diagnoseExport(await openZipArchive(readFixtureArchive('extended-malformed')))
    expect(files).toEqual([
      {
        path: 'Spotify Extended Streaming History/Streaming_History_Audio_2024-2025_0.json',
        bytes: expect.any(Number),
        rows: 2,
        headers: expect.arrayContaining(['ts', 'ms_played', 'spotify_track_uri']),
      },
      {
        path: 'Spotify Extended Streaming History/Streaming_History_Audio_2025-2026_1.json',
        bytes: expect.any(Number),
        rows: null,
        headers: null,
      },
    ])
  })

  it('does not read a library file travelling with an extended package', async () => {
    const archive = recordingArchive(await openZipArchive(readFixtureArchive('extended-nested-folder')))
    const { files } = await diagnoseExport(archive)
    const library = files.find((file) => file.path.toLowerCase().endsWith('yourlibrary.json'))!
    expect(library).toMatchObject({ rows: null, headers: null })
    expect(archive.readPaths).not.toContain(library.path)
    expect(archive.readPaths.length).toBeGreaterThan(0)
  })
})

File: client/test/import/diagnostics_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mixtape/import/diagnostics.dart';
import 'package:mixtape/import/zip_reader.dart';

import 'fixtures.dart';

void main() {
  test('lists every file with bytes; rows and headers only for files the parser reads', () async {
    final archive = ZipExportArchive.open(fixtureArchive('account-pii-present'));
    try {
      final diagnostics = await diagnoseExport(archive);
      expect(diagnostics.source, 'spotify_export');
      expect(diagnostics.parserVersion, isNotEmpty);

      final byPath = {for (final f in diagnostics.files) f.path: f};
      expect(byPath.keys.toList(), [
        'Spotify Account Data/Identity.json',
        'Spotify Account Data/Inferences.json',
        'Spotify Account Data/Payments.json',
        'Spotify Account Data/Playlist1.json',
        'Spotify Account Data/Userdata.json',
        'Spotify Account Data/YourLibrary.json',
        '__MACOSX/._YourLibrary.json',
      ]);

      for (final sentinel in [
        'Spotify Account Data/Identity.json',
        'Spotify Account Data/Inferences.json',
        'Spotify Account Data/Payments.json',
        'Spotify Account Data/Userdata.json',
        '__MACOSX/._YourLibrary.json',
      ]) {
        final file = byPath[sentinel]!;
        expect(file.rows, isNull, reason: sentinel);
        expect(file.headers, isNull, reason: sentinel);
        expect(file.bytes, greaterThanOrEqualTo(2048), reason: sentinel);
      }
      expect(byPath['Spotify Account Data/Identity.json']!.bytes, 2068);

      final playlist = byPath['Spotify Account Data/Playlist1.json']!;
      expect(playlist.rows, 1);
      expect(playlist.headers, ['playlists']);
      final library = byPath['Spotify Account Data/YourLibrary.json']!;
      expect(library.rows, 2);
      expect(library.headers, contains('tracks'));
      expect(library.headers, contains('artists'));

      final text = diagnostics.canonicalJsonString();
      expect(text, isNot(contains('DO-NOT-READ')));
      expect(text, isNot(contains('spotify:')));
      // No value from the read files leaks: only keys are reported.
      final definition = caseDefinition(fixtureCases().singleWhere((d) => d.path.endsWith('account-pii-present')));
      for (final value in _leafStrings(definition['entries'])) {
        expect(text, isNot(contains(value)), reason: 'value leaked');
      }
    } finally {
      await archive.close();
    }
  });

  test('history headers come from the first array element and a broken file has null rows', () async {
    final archive = ZipExportArchive.open(fixtureArchive('extended-malformed'));
    try {
      final diagnostics = await diagnoseExport(archive);
      expect(diagnostics.files, hasLength(2));
      final fine = diagnostics.files.first;
      expect(fine.rows, 2);
      expect(fine.headers, contains('ts'));
      expect(fine.headers, contains('ms_played'));
      final broken = diagnostics.files.last;
      expect(broken.rows, isNull);
      expect(broken.headers, isNull);
      expect(broken.bytes, greaterThan(0));
      final json = diagnostics.toCanonicalJson();
      expect(json.keys.toList(), ['source', 'files', 'parserVersion']);
      expect((json['files'] as List).first, {
        'path': fine.path,
        'bytes': fine.bytes,
        'rows': 2,
        'headers': fine.headers,
      });
    } finally {
      await archive.close();
    }
  });
}

Iterable<String> _leafStrings(Object? node) sync* {
  if (node is Map) {
    for (final entry in node.entries) {
      // Entry paths are reported by design; bodies are not.
      if (entry.key == 'path') continue;
      yield* _leafStrings(entry.value);
    }
  } else if (node is List) {
    for (final item in node) {
      yield* _leafStrings(item);
    }
  } else if (node is String && node.length >= 4) {
    yield node;
  }
}
